src/ast/SemanticChecksVisitor.java

```diff
@@ -35,32 +42,294 @@ public void visit(Program program) {
+        return false;
+    }
+
+    /**
+     * This method checks if a given expression is of type boolean
+     * For an illegal and expressions it returns false
+     * @param e
+     * @return
+     */
+    private boolean isBooleanExpr(Expr e){
+        String dinamicExprName = e.getClass().getName();
+        if(dinamicExprName.equals("ast.TrueExpr") || dinamicExprName.equals("ast.FalseExpr")){
+            return true;
+        }
+
+        if(dinamicExprName.equals("ast.AndExpr")){
+            AndExpr ae = (AndExpr)e;
+            //both need to be boolean
+            return isBooleanExpr(ae.e1()) && isBooleanExpr(ae.e2());
+        }
+
+        if(dinamicExprName.equals("ast.NotExpr")){
+            return isBooleanExpr(((NotExpr)e).e());
+        }
+        //if we're here than the Expr doesn't fit a boolean type
+        return false;
+    }
+
+    /**
+     * given a MethodCallExpr this method returns the AstType that method returns
+     * if it's a reference type this method makes sure to have the correct id set
+     * @param e
+     * @return AstType that method called returns, null if the call is illegal
+     */
+    private AstType methodCallerToReturnType(MethodCallExpr e){
+        String ownerExprType = e.ownerExpr().getClass().getName();
+        //we'll get the invoker class name and than lookup the vtable to get the MethodDecl
+        String callerClassName;
+
+        if(ownerExprType.equals("ast.ThisExpr")){
+            callerClassName = e.enclosingScope().getParent().scopeName();
+        }
+
+        else if(ownerExprType.equals("ast.NewObjectExpr")){
+            callerClassName = ((NewObjectExpr)(e.ownerExpr())).classId();
+        }
 
+        else if(ownerExprType.equals("ast.IdentifierExpr")){
+            IdentifierExpr ie = (IdentifierExpr)(e.ownerExpr());
+            //get declaration node of invoker
+            AstNode decl = STLookup.getDeclNode(STLookup.findDeclTable(ie.id(),forest,e.enclosingScope(),programST), ie.id());
+            VariableIntroduction varIntro = (VariableIntroduction)decl;
+            //check to see that caller is RefType
+            if(!varIntro.type().getClass().getName().equals("ast.RefType")) return null;
+            callerClassName = ((RefType)varIntro.type()).id();
+        }
+        /* if the caller is none of the above than it's not a legal call
+         * check https://www.cs.tau.ac.il/research/yotam.feldman/courses/wcc20/project.html
+         * two bullets before last in the overview to be convinced of this*/
+        else return null;
+
+        //check to see that owner is actually a class defined in the program
+        if(forest.nameToClassDecl(callerClassName) == null) return null;
+
+        for(STSymbol symbol : vtables.get(callerClassName)){
+            if(symbol.name().equals(e.methodId())){
+                return ((MethodDecl)(symbol.declaration())).returnType();
+            }
+        }
+        //if we're here than the method called doesn't exists
```

Comment:
shouldn't we call it a day and say the visit has failed?

Reply:
In a sense yes.
But what I did was make the assumption that if the return value is null than the caller to this method will take care of it